### PR TITLE
Feat: add workflow record actions

### PR DESCRIPTION
**Pull request by FogDong**
### Description of your changes

Fixes #

I have:

- [ ] Read and followed KubeVela's [contribution process](https://github.com/oam-dev/kubevela/blob/master/contribute/create-pull-request.md).
- [ ] [Related Docs](https://github.com/oam-dev/kubevela.io) updated properly. In a new feature or configuration option, an update to the documentation is necessary. 
- [ ] Run `make reviewable` to ensure this PR is ready for review.
- [ ] Added `backport release-x.y` labels to auto-backport this PR if necessary.

### CODE REVIEW

**Review by hongchaodeng — APPROVED** [review]

File: pkg/apiserver/rest/usecase/workflow.go [hunk 1]
```diff
@@ -288,115 +291,251 @@ func (w *workflowUsecaseImpl) DetailWorkflowRecord(ctx context.Context, workflow
...
+	return nil
+}
+
+func (w *workflowUsecaseImpl) RollbackRecord(ctx context.Context, appModel *model.Application, recordName, revisionName string) error {
+	oamApp, err := w.checkRecordSuspended(ctx, appModel)
+	if err != nil {
+		return err
+	}
+
+	var record = &model.WorkflowRecord{
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/workflow.go:459]: AppPrimaryKey+Name == recordPrimaryKey, please change PrimaryKey function of WorkflowRecord.

File: pkg/apiserver/rest/usecase/workflow.go [hunk 2]
```diff
@@ -288,115 +291,251 @@ func (w *workflowUsecaseImpl) DetailWorkflowRecord(ctx context.Context, workflow
...
+	if err := yaml.Unmarshal([]byte(rollbackRevision.ApplyAppConfig), rollBackApp); err != nil {
+		return err
+	}
+	// replace the application spec
+	oamApp.Spec.Components = rollBackApp.Spec.Components
+	if oamApp.Annotations == nil {
+		oamApp.Annotations = make(map[string]string)
+	}
+	newRecordName := utils.GenerateVersion(record.WorkflowPrimaryKey)
+	oamApp.Annotations[oam.AnnotationDeployVersion] = rollbackRevision.Version
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/workflow.go:498]: What is the diff between AnnotationDeployVersion and AnnotationPublishVersion?

File: pkg/apiserver/rest/usecase/workflow.go [hunk 3]
```diff
@@ -288,115 +291,251 @@ func (w *workflowUsecaseImpl) DetailWorkflowRecord(ctx context.Context, workflow
...
+	oamApp.Annotations[oam.AnnotationPublishVersion] = newRecordName
+
+	// create a new workflow record
+	if err := w.CreateWorkflowRecord(ctx, oamApp, revisionName); err != nil {
+		return err
+	}
+
+	if err := w.kubeClient.Update(ctx, oamApp); err != nil {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/workflow.go:506]: Is it better to use `apply.Applicator`?

File: pkg/apiserver/rest/usecase/workflow.go [hunk 4]
```diff
@@ -288,115 +291,251 @@ func (w *workflowUsecaseImpl) DetailWorkflowRecord(ctx context.Context, workflow
...
+		AppPrimaryKey: appModel.Name,
+		Name:          revisionName,
+	}
+	if err := w.ds.Get(ctx, &rollbackRevision); err != nil {
+		return err
+	}
+
+	originalVersion := originalRevision.Version
+	// update the original revision's version
+	originalRevision.Version = rollbackRevision.Version
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/workflow.go:483]: This should not be updated.

File: pkg/apiserver/rest/usecase/workflow.go [hunk 5]
```diff
@@ -288,115 +296,256 @@ func (w *workflowUsecaseImpl) DetailWorkflowRecord(ctx context.Context, workflow
...
+		}
+		return err
+	}
+
+	return nil
+}
+
+func (w *workflowUsecaseImpl) checkRecordRunning(ctx context.Context, appModel *model.Application) (*v1beta1.Application, error) {
+	oamApp := &v1beta1.Application{}
+	if err := w.kubeClient.Get(ctx, types.NamespacedName{Name: appModel.Name, Namespace: appModel.Namespace}, oamApp); err != nil {
```
Comment (barnettZQG) [inline pkg/apiserver/rest/usecase/workflow.go:532]: 使用converAppName方法计算实际名称。

File: pkg/apiserver/model/application.go [hunk 6]
```diff
@@ -216,7 +216,7 @@ func (a *ApplicationRevision) TableName() string {
 
 // PrimaryKey return custom primary key
 func (a *ApplicationRevision) PrimaryKey() string {
-	return fmt.Sprintf("%s-%s", a.AppPrimaryKey, a.Version)
+	return a.Version
```
Comment (barnettZQG) [inline pkg/apiserver/model/application.go:219]:
```suggestion
	return fmt.Sprintf("%s-%s", a.AppPrimaryKey, a.Version)
```

**Pull request merged**